### ISSUE TITLE
Do not use final classes

### ISSUE THREAD

**Issue by SerafimArts** [cycle/orm#12]
Objective: To adapt the cycle to the existing solutions (Symfony, Laravel, е.g.)

The easiest implementation:

```php
class MyDatabaseManager extends Spiral\Database\DatabaseManager 
{
    protected function makeDatabase(DatabasePartial $database): Database
    {
        return new Database (
            $database->getName(),
            $database->getPrefix(),
            // In this place we can create our own driver using the existing connection in the framework, like:
            // DoctrineMySQLDriver::fromPdo($this->symfony->connection($database->getName())->getPdo())
    }
}

class DoctrineMySQLDriver extends MySQLDriver { ... }
```

But such an implementation is impossible, we cannot reuse the existing connection due to the presence of the final classes in the code.

DatabaseManager doesn't even have an interface so we can completely rewrite it, and the code uses impl: https://github.com/cycle/orm/blob/master/src/Factory.php#L42 

Proposal:
1) Use interfaces at "interaction" points, like:
```php
interface DatabaseManagerInterface 
{
    public function getDatabases(): iterable;
    public function database(string $database = null): Database;
    public function addDatabase(Database $database);
}
```
2) Delete and do not use the `final` keyword.
3) Other ideas how to adapt the configuration of Doctrine or Eloquent to reuse it in a cycle?

**Comment by wolfy-j** [comment]
Yes, DM interface will be added. All Databases support DriverInterface, so you can make your implementation if you want. The only trick will be to adapt schema introspection classes, but they do not used in runtime.

**Comment by wolfy-j** [comment]
I guess the biggest problem here is final implementation of drivers. I’ll work to make bridge easier.

**Comment by SerafimArts** [comment]
This is the only place where the manager is used: https://github.com/cycle/orm/blob/master/src/Factory.php#L128-L131

It turns out we should do this:

```php
namespace Spiral\Database;

interface ProvidesDatabase 
{
     public function database(string $database): DatabaseInterface;
}
```

And change impl to this:
```php
namespace Spiral\Database;

class DatabaseManager implements ProvidesDatabase { ... }
```

```php
namespace Cycle\ORM;

class Factory implements ProvidesDatabase 
{
    public function __construct(ProvidesDatabase $dbm, ...) { ... } // Change signature here to ProvidesDatabase interface
}
```

**Comment by wolfy-j** [comment]
Yes, this sounds about right.

**Comment by wolfy-j** [comment]
Fixed in **v1.0.4**. For consustency, I've used Interface suffix.
